src/aurora/erfwriter.cpp

```diff
@@ -0,0 +1,108 @@
+ * xoreos is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 3
+ * of the License, or (at your option) any later version.
+ *
+ * xoreos is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with xoreos. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/** @file
+ *  Writing BioWare's ERFs (encapsulated resource file).
+ */
+
+#include <ctime>
+#include "erfwriter.h"
+
+namespace Aurora {
+
+ERFWriter::ERFWriter(uint32 id, uint32 fileCount, Common::SeekableWriteStream &stream, LocString description) :
+		_stream(stream), _id(id), _currentFileCount(0), _fileCount(fileCount), _description(description) {
+	stream.writeUint32BE(_id);
+	stream.writeUint32BE(MKTAG('V', '1', '.', '0'));
+
+	// Write Header.
+	stream.writeUint32LE(_description.getNumStrings()); // Language count.
+	stream.writeUint32LE(_description.getWrittenSize()); // Localized string size.
+
+	stream.writeUint32LE(_fileCount); // Entry Count.
+
+	uint32 locStringTableOffset = 160; // 160 is the size of the erf header.
+	_keyTableOffset = locStringTableOffset + _description.getWrittenSize();
+	_resourceTableOffset = _keyTableOffset + _fileCount * 24;
+
+	stream.writeUint32LE(locStringTableOffset); // LocString offset.
+	stream.writeUint32LE(_keyTableOffset); // Key List offset.
+	stream.writeUint32LE(_resourceTableOffset); // Resource offset.
+
+	// Write the creation time of the file.
+	std::time_t now = std::time(0);
+	std::tm *timepoint = std::localtime(&now);
+	stream.writeUint32LE(timepoint->tm_year);
+	stream.writeUint32LE(timepoint->tm_yday);
+
+	// Write the description string reference.
+	if (_description.getNumStrings())
+		stream.writeUint32LE(_description.getID());
+	else
+		stream.writeUint32LE(0);
+
+	// Write 116 bytes of reserved header data.
+	stream.writeZeros(116);
+
+	// Write the Localized string table.
+	_description.writeLocString(stream);
+
+	// Write the empty key list.
+	stream.writeZeros(_fileCount * 24);
+
+	// The offset to the resource table plus the size of the source table
+	_offsetToResourceData = _resourceTableOffset + 8 * _fileCount;
+	// Write the empty resource list.
+	stream.writeZeros(8 * _fileCount);
+}
+
+ERFWriter::~ERFWriter() {
+}
+
+void ERFWriter::add(const Common::UString resRef, uint16 resType, unsigned int size, Common::ReadStream &stream) {
+	if (_currentFileCount == _fileCount)
+		throw Common::Exception("More files added than expected");
+
+	// Write the key table entry
+	_stream.seek(_keyTableOffset + _currentFileCount * 24);
+
+	char cname[16];
+	memset(cname, '\0', 16);
+	memcpy(cname, resRef.c_str(), MIN(static_cast<size_t>(16), resRef.size()));
```

Comment:
`MIN<size_t>()`.

Is the copying necessary, though? You could just write `MIN<size_t>(16, resRef.size())` from `resRef.c_str()` and then `writeZeros(16 - MIN<size_t>(16, resRef.size())`.